### PR TITLE
GUI: Import wallet imports addressbook too

### LINKED ISSUE #565: Recovery procedure doesn't restore a saved address book
#### System information
Semux/v1.0.0-4a83a2d/Windows/amd64

OS & Version: Windows 7x64

#### Expected behaviour
Wallet backup contains an address book. "Restore wallet" procedure recoveres the saved address book.

#### Actual behaviour
A wallet file stores an address book, but recovery procedure doesn't restore a saved address book.

#### Steps to reproduce the behaviour
Restore a wallet with an addressbook from a backup. The address book is empty (isn't recovered).

Bounty address is 0x9f0b3262b87d5350b81da69a8f79fd7893f9eff9

### PR DESCRIPTION
**Pull request by orogvany**
Import the new wallets address book.
Chose to not override any existing aliases, but could
go either way on that one.  There is little chance of overlap.

Prefer to keep existing wallet as source of truth if conflicts arise

fixes #565

### CODE REVIEW

File: src/main/java/org/semux/gui/MenuBar.java [hunk 1]
```diff
@@ -182,6 +184,17 @@ protected void recoverAccounts() {
 
                 Wallet wallet = gui.getKernel().getWallet();
                 int n = wallet.addAccounts(w.getAccounts());
+                // add address book entries
+                Map<ByteArray, String> importedAliases = w.getAddressAliases();
+                for (Map.Entry<ByteArray, String> alias : importedAliases.entrySet()) {
+                    byte[] address = alias.getKey().getData();
+
+                    // don't override existing wallet's aliases.
+                    if (wallet.getAddressAlias(address) == null) {
```
Comment (cryptokat) [inline src/main/java/org/semux/gui/MenuBar.java]: This needs to be changed to `!wallet.getAddressAlias(address).isPresent()`

**Comment by orogvany** [comment]
I'll extract the wallet import code, so both places can use same importing.  Was not aware that it was done separately 2 different places

File: src/main/java/org/semux/gui/MenuBar.java (hunk 1 above)
Comment (orogvany) [reply src/main/java/org/semux/gui/MenuBar.java]: done, thanks

File: src/main/java/org/semux/gui/WelcomeFrame.java [hunk 2]
```diff
@@ -283,7 +283,7 @@ protected void ok() {
                 JOptionPane.showMessageDialog(this, GuiMessages.get("NoAccountFound"));
             } else {
                 if (wallet.unlock(password)
-                        && wallet.addAccounts(w.getAccounts()) > 0
+                        && wallet.addWallet(w) > 0
                         && wallet.flush()) {
```
Comment (cryptokat) [inline src/main/java/org/semux/gui/WelcomeFrame.java]: The wallet flushes 2 times here.

File: src/main/java/org/semux/core/Wallet.java [hunk 3]
```diff
@@ -515,4 +515,31 @@ private void requireUnlocked() throws WalletLockedException {
...
+        for (Map.Entry<ByteArray, String> alias : importedAliases.entrySet()) {
+            byte[] address = alias.getKey().getData();
+
+            // don't override existing wallet's aliases.
+            if (!getAddressAlias(address).isPresent()) {
+                setAddressAlias(address, alias.getValue());
+            }
+        }
+
+        flush();
```
Comment (cryptokat) [inline src/main/java/org/semux/core/Wallet.java]: It would provide better UX if we report an error when the wallet is failed to save on disk.

Comment (orogvany) [reply src/main/java/org/semux/core/Wallet.java]: done.

File: src/main/java/org/semux/gui/WelcomeFrame.java (hunk 2 above)
Comment (orogvany) [reply src/main/java/org/semux/gui/WelcomeFrame.java]: You're right.  The addWallet method should not call flush() for consistency with other wallet modifying methods.  This has been corrected.

Comment (orogvany) [reply src/main/java/org/semux/gui/WelcomeFrame.java]: Not sure how to resolve this change request?

**Pull request merged**